### PR TITLE
Grainsize sundials

### PR DESCRIPTION
**Pull request by jdannberg**
This PR changes the grain size material model to use SUNDIALs to compute the grain size change. 
This is still about 30% slower than the original implementation. I am also not sure how to set all the ARKode parameters. The SUNDIALs documentation says that it tries 7 times to decrease the timestep if the error is too large, and that was not enough in some of the cases I tried, so I'd like to increase this number (rather than decreasing the initial step size). 

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### CODE REVIEW

**Comment by bangerth** [comment]
I will take a look on occasion. Would you mind rebasing to remove the merge conflict?

**Comment by jdannberg** [comment]
I rebased to the current main, and I figured out how to properly use the tolerance of ARKode. I also added a test case that does not depend on resolution (the velocity is zero and it only runs in one cell) and has an analytical solution, to be better able to judge the accuracy. With these changes, the new version with SUNDIALs is more accurate, and it is also faster. The more the grain size changes within a time step, the more efficient SUNDIALs is compared to the old implementation. 

Here are some numbers from my laptop, run with the new test case I added, and changing the global refinements to 7 (just so that everything runs on lots of points and takes longer). I made 3 different test cases, one with the original reaction rate, one with 10x faster reactions, and one with 100x faster reaction (by changing the Grain growth rate constant). 
| | Total wallclock time elapsed since start | Assemble composition system |
|---|---|---|
|original growth rate |
|new | 6.07s | 1.6s |
|old | 8.01s | 2.48s |
|10x growth rate | | |
|new | 7.57s | 2.4s | 
|old | 13.7s | 5.24s | 
|100x growth rate | 
|new | 7.4s | 2.27s | 
|old | 17.5s | 7.11s |

**Comment by bangerth** [comment]
That's a nice improvement!

File: include/aspect/material_model/grain_size.h [hunk 1]
```diff
@@ -192,6 +197,9 @@ namespace aspect
         std::vector<double> reciprocal_required_strain;
         std::vector<double> recrystallized_grain_size;
 
+        SUNDIALS::ARKode<VectorType>::AdditionalData data;
+        mutable VectorType grain_sizes;
```
Comment (bangerth) [inline include/aspect/material_model/grain_size.h:201]: Do these two variables need to be class members? I think you could probably just move those into the function where you end up using those, and that makes the function thread-safe too.

File: source/material_model/grain_size.cc [hunk 2]
```diff
@@ -110,6 +110,25 @@ namespace aspect
...
+      data.initial_step_size = this->get_timestep();
+      data.output_period = this->get_timestep();
+      data.minimum_step_size = 1.e-6*this->get_timestep();
+      data.maximum_order = 3;
+      data.maximum_non_linear_iterations = 30;
+
+      // Because both tolerances are added, we set the absolute
+      // tolerance to 0.
+      data.relative_tolerance = 1e-3;
+      data.absolute_tolerance = 0;
```
Comment (bangerth) [inline source/material_model/grain_size.cc:128]: I would move all of this right above the place where you now create the ARKode solver object.

File: source/material_model/grain_size.cc [hunk 3]
```diff
@@ -222,165 +241,187 @@ namespace aspect
...
-                                           this->get_adiabatic_conditions().temperature(position)
-                                           :
-                                           temperature;
-
-      do
+      std::vector<double> grain_size_changes (in.n_evaluation_points());
+      std::vector<std::vector<double>> reaction_terms (in.n_evaluation_points(), std::vector<double>(this->n_compositional_fields(), 0.0));
+
+      // Set up an integer that tells us which phase transition has been crossed inside of the cell.
+      std::vector<int> crossed_transitions (in.n_evaluation_points(), -1);
```
Comment (bangerth) [inline source/material_model/grain_size.cc:256]: Probably need to update the comment for the fact that you now have a vector.

File: source/material_model/grain_size.cc [hunk 4]
```diff
@@ -222,165 +241,187 @@ namespace aspect
...
+      const double timestep = this->simulator_is_past_initialization()
+                              ?
+                              this->get_timestep()
+                              :
+                              0.0;
+
+      for (unsigned int i=0; i<in.n_evaluation_points(); ++i)
         {
-          time += grain_growth_timestep;
+          grain_sizes[i]=in.composition[i][grain_size_index];
```
Comment (bangerth) [inline source/material_model/grain_size.cc:268]:
```suggestion
          grain_sizes[i] = in.composition[i][grain_size_index];
```

File: source/material_model/grain_size.cc [hunk 5]
```diff
@@ -222,165 +241,187 @@ namespace aspect
...
 
-          if (timestep - time < 0)
-            {
-              grain_growth_timestep = timestep - (time - grain_growth_timestep);
-              time = timestep;
-            }
+          if ((in.composition[i][grain_size_index] != in.composition[i][grain_size_index]) || this->get_timestep() == 0.0
+              || in.composition[i][grain_size_index] < std::numeric_limits<double>::min())
+            return reaction_terms;
```
Comment (bangerth) [inline source/material_model/grain_size.cc:272]: I don't understand this. The function is supposed to return grain size changes (based on its name) but returns reaction rates here. This cannot be correct based on units alone. A grain size change has units meters, the reation rate has units meters per second.

Separately, did you mean to exit out of the function if *any* of the points satisfy this condition? What happens to the other points?

File: source/material_model/grain_size.cc [hunk 6]
```diff
@@ -222,165 +241,187 @@ namespace aspect
...
+        }
+
+      SUNDIALS::ARKode<VectorType> ode(data);
```
Comment (bangerth) [inline source/material_model/grain_size.cc:275]: Put the declaration and initialization of the `data` object right above here.

**Review by bangerth — COMMENTED** [review]
Looks pretty nice!

File: include/aspect/material_model/grain_size.h (hunk 1 above)
Comment (jdannberg) [reply include/aspect/material_model/grain_size.h:201]: I first had this in the function, but @gassmoeller suggested to move them out to make it faster (so we do not have to create them again for every cell).

File: source/material_model/grain_size.cc (hunk 5 above)
Comment (jdannberg) [reply source/material_model/grain_size.cc:272]: The reaction terms are not reaction rates, they are changes in the compositional fields. That is the variable name we use in all material models.

**Comment by jdannberg** [comment]
I have addressed your comments and updated the test results. I went through all the tests one by one, and the new results all make sense, except for the grain_size_growth_log.prm test. This one clearly has wrong test results, but they were also wrong before, so it is not related to this change. This was an option we introduced before we could advect grain size on particles, so now it does not make sense to use it anymore. Therefore I would just completely remove that option in a follow-up PR.

**Comment by bangerth** [comment]
OK, tell me when you want me to look again!

**Comment by jdannberg** [comment]
@bangerth Can you have another look?

The last test that was not working was a cookbook, so it was an application model using grain size. The problem was that it was crashing because the initial step size passed into ARKode was too long. So I made this step size smaller, and I made sure that it does not affect the timing in the test case (ARKode increases the step size by a factor of 20 if it’s too small, so it would do at most 2 steps more if this initial step size was too small). However, how to pick this parameter likely depends on the model, so I’d like to make this an input parameter in a follow-up PR. 

The test output also changed, so I ran the whole cookbook again with both the old and new version, and with different time step sizes. The figures below shows that the version with ARKode is the more accurate one (it is closer to the version with the small time steps). So we know why the test changed and that the new version is right. 

time_step_convergence

The new version also takes way longer for this application case (grain size is tracked on particles, and Particles: Update properties can take 10x as long with the new ARKode implementation). But maybe that's something we can try to fix as a follow-up as well.

**Comment by jdannberg** [comment]
Update: ARKode is slower in models with particles (as my application example) because they call the evaluate function once per particle, not per cell. In very simple problems, this can mean that the ARKode implementation is slower than the old one. However in my simple test cases the slowdown was never as dramatic (no matter what the grain growth rate was, it was at most 20%), so the dramatic slowdown in the application model must have been caused by something else.

**Comment by bangerth** [comment]
Can you squash/fixup commits?

File: include/aspect/material_model/grain_size.h [hunk 7]
```diff
@@ -36,6 +37,7 @@ namespace aspect
   namespace MaterialModel
   {
     using namespace dealii;
+    using VectorType = Vector<double>;
```
Comment (bangerth) [inline include/aspect/material_model/grain_size.h:40]: I don't think you use this anywhere outside the one function where you call ARKode. Can you move that there (i.e., into the function)?

**Review by bangerth — COMMENTED** [review]
Just this one issue. Thanks for all the careful testing!

**Comment by jdannberg** [comment]
I have squashed my commits and moved the `using VectorType = Vector<double>`. I have also added a changelog entry (since I had forgotten to do that).

**Review by bangerth — APPROVED** [review]

**Pull request merged**